# Рекомендации на основе содержания
Использовать датасет мувилен.

Построить рекомендации (регрессия, предсказываем оценку) на фичах:
- tf-idf на тегах и жанрах
- средние оценки (+median, variance и др.) пользователя и фильма

Оценить RMSE на тестовой выборке

In [69]:
import pandas as pd
import numpy as np
from collections import Counter
from datetime import datetime
from tqdm.notebook import tqdm
import matplotlib.pyplot as plt

In [70]:
links = pd.read_csv('links.csv')
movies = pd.read_csv('movies.csv')
ratings = pd.read_csv('ratings.csv')
tags = pd.read_csv('tags.csv')

In [71]:
movies

movieId                                      title  \
0           1                           Toy Story (1995)   
1           2                             Jumanji (1995)   
2           3                    Grumpier Old Men (1995)   
3           4                   Waiting to Exhale (1995)   
4           5         Father of the Bride Part II (1995)   
...       ...                                        ...   
9737   193581  Black Butler: Book of the Atlantic (2017)   
9738   193583               No Game No Life: Zero (2017)   
9739   193585                               Flint (2017)   
9740   193587        Bungo Stray Dogs: Dead Apple (2018)   
9741   193609        Andrew Dice Clay: Dice Rules (1991)   

                                           genres  
0     Adventure|Animation|Children|Comedy|Fantasy  
1                      Adventure|Children|Fantasy  
2                                  Comedy|Romance  
3                            Comedy|Drama|Romance  
4                                          Comedy  
...                                           ...  
9737              Action|Animation|Comedy|Fantasy  
9738                     Animation|Comedy|Fantasy  
9739                                        Drama  
9740                             Action|Animation  
9741                                       Comedy  

[9742 rows x 3 columns]

Делаем вектор из жанров:

In [72]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.neighbors import NearestNeighbors

In [73]:
movies.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [74]:
def change_string(s):
    return ' '.join(s.replace(' ', '').replace('-', '').split('|'))

In [75]:
movie_genres = [change_string(g) for g in movies.genres.values]
movie_genres[:10]

['Adventure Animation Children Comedy Fantasy',
 'Adventure Children Fantasy',
 'Comedy Romance',
 'Comedy Drama Romance',
 'Comedy',
 'Action Crime Thriller',
 'Comedy Romance',
 'Adventure Children',
 'Action',
 'Action Adventure Thriller']

In [76]:
tfidf = TfidfVectorizer()
X_train_tfidf = tfidf.fit_transform(movie_genres)
X_train_tfidf

<9742x20 sparse matrix of type '<class 'numpy.float64'>'
	with 22084 stored elements in Compressed Sparse Row format>

In [77]:
print(X_train_tfidf)

  (0, 8)	0.482990142708577
  (0, 4)	0.26758647689140014
  (0, 3)	0.5048454681396087
  (0, 2)	0.5162254711770092
  (0, 1)	0.41684567364693936
  (1, 8)	0.5936619434123594
  (1, 3)	0.620525172745643
  (1, 1)	0.5123612074824268
  (2, 15)	0.8210088907493954
  (2, 4)	0.5709154064399099
  (3, 7)	0.46640480307738325
  (3, 15)	0.726240982959826
  (3, 4)	0.5050154397005037
  (4, 4)	1.0
  (5, 17)	0.5420423542868653
  (5, 5)	0.6359470441562756
  (5, 0)	0.5493281743985542
  (6, 15)	0.8210088907493954
  (6, 4)	0.5709154064399099
  (7, 3)	0.7711121633813997
  (7, 1)	0.6366993258087036
  (8, 0)	1.0
  (9, 17)	0.5457299419583338
  (9, 0)	0.5530653284926609
  (9, 1)	0.6295217016667962
  :	:
  (9731, 16)	0.5912169125932074
  (9731, 0)	0.47946635727353987
  (9731, 4)	0.3503331832902893
  (9731, 1)	0.5457483258721835
  (9732, 16)	0.5491808953132112
  (9732, 0)	0.4453758980017919
  (9732, 4)	0.32542420076143624
  (9732, 2)	0.6278054979536716
  (9733, 7)	0.43179372819853856
  (9733, 2)	0.9019723811117537
  (9

In [78]:
movie_genres_table = pd.DataFrame(X_train_tfidf.toarray(), columns=tfidf.get_feature_names_out())
movie_genres_table

action  adventure  animation  children    comedy  crime  documentary  \
0     0.000000   0.416846   0.516225  0.504845  0.267586    0.0          0.0   
1     0.000000   0.512361   0.000000  0.620525  0.000000    0.0          0.0   
2     0.000000   0.000000   0.000000  0.000000  0.570915    0.0          0.0   
3     0.000000   0.000000   0.000000  0.000000  0.505015    0.0          0.0   
4     0.000000   0.000000   0.000000  0.000000  1.000000    0.0          0.0   
...        ...        ...        ...       ...       ...    ...          ...   
9737  0.436010   0.000000   0.614603  0.000000  0.318581    0.0          0.0   
9738  0.000000   0.000000   0.682937  0.000000  0.354002    0.0          0.0   
9739  0.000000   0.000000   0.000000  0.000000  0.000000    0.0          0.0   
9740  0.578606   0.000000   0.815607  0.000000  0.000000    0.0          0.0   
9741  0.000000   0.000000   0.000000  0.000000  1.000000    0.0          0.0   

         drama   fantasy  filmnoir  horror  imax  musical  mystery  \
0     0.000000  0.482990       0.0     0.0   0.0      0.0      0.0   
1     0.000000  0.593662       0.0     0.0   0.0      0.0      0.0   
2     0.000000  0.000000       0.0     0.0   0.0      0.0      0.0   
3     0.466405  0.000000       0.0     0.0   0.0      0.0      0.0   
4     0.000000  0.000000       0.0     0.0   0.0      0.0      0.0   
...        ...       ...       ...     ...   ...      ...      ...   
9737  0.000000  0.575034       0.0     0.0   0.0      0.0      0.0   
9738  0.000000  0.638968       0.0     0.0   0.0      0.0      0.0   
9739  1.000000  0.000000       0.0     0.0   0.0      0.0      0.0   
9740  0.000000  0.000000       0.0     0.0   0.0      0.0      0.0   
9741  0.000000  0.000000       0.0     0.0   0.0      0.0      0.0   

      nogenreslisted   romance  scifi  thriller  war  western  
0                0.0  0.000000    0.0       0.0  0.0      0.0  
1                0.0  0.000000    0.0       0.0  0.0      0.0  
2                0.0  0.821009    0.0       0.0  0.0      0.0  
3                0.0  0.726241    0.0       0.0  0.0      0.0  
4                0.0  0.000000    0.0       0.0  0.0      0.0  
...              ...       ...    ...       ...  ...      ...  
9737             0.0  0.000000    0.0       0.0  0.0      0.0  
9738             0.0  0.000000    0.0       0.0  0.0      0.0  
9739             0.0  0.000000    0.0       0.0  0.0      0.0  
9740             0.0  0.000000    0.0       0.0  0.0      0.0  
9741             0.0  0.000000    0.0       0.0  0.0      0.0  

[9742 rows x 20 columns]

In [79]:
movies

movieId                                      title  \
0           1                           Toy Story (1995)   
1           2                             Jumanji (1995)   
2           3                    Grumpier Old Men (1995)   
3           4                   Waiting to Exhale (1995)   
4           5         Father of the Bride Part II (1995)   
...       ...                                        ...   
9737   193581  Black Butler: Book of the Atlantic (2017)   
9738   193583               No Game No Life: Zero (2017)   
9739   193585                               Flint (2017)   
9740   193587        Bungo Stray Dogs: Dead Apple (2018)   
9741   193609        Andrew Dice Clay: Dice Rules (1991)   

                                           genres  
0     Adventure|Animation|Children|Comedy|Fantasy  
1                      Adventure|Children|Fantasy  
2                                  Comedy|Romance  
3                            Comedy|Drama|Romance  
4                                          Comedy  
...                                           ...  
9737              Action|Animation|Comedy|Fantasy  
9738                     Animation|Comedy|Fantasy  
9739                                        Drama  
9740                             Action|Animation  
9741                                       Comedy  

[9742 rows x 3 columns]

Прилепим обратно айди:

In [81]:
movie_genres_table_id = movie_genres_table.join(
    movies['movieId'],
    how='inner')

In [82]:
movie_genres_table_id

action  adventure  animation  children    comedy  crime  documentary  \
0     0.000000   0.416846   0.516225  0.504845  0.267586    0.0          0.0   
1     0.000000   0.512361   0.000000  0.620525  0.000000    0.0          0.0   
2     0.000000   0.000000   0.000000  0.000000  0.570915    0.0          0.0   
3     0.000000   0.000000   0.000000  0.000000  0.505015    0.0          0.0   
4     0.000000   0.000000   0.000000  0.000000  1.000000    0.0          0.0   
...        ...        ...        ...       ...       ...    ...          ...   
9737  0.436010   0.000000   0.614603  0.000000  0.318581    0.0          0.0   
9738  0.000000   0.000000   0.682937  0.000000  0.354002    0.0          0.0   
9739  0.000000   0.000000   0.000000  0.000000  0.000000    0.0          0.0   
9740  0.578606   0.000000   0.815607  0.000000  0.000000    0.0          0.0   
9741  0.000000   0.000000   0.000000  0.000000  1.000000    0.0          0.0   

         drama   fantasy  filmnoir  ...  imax  musical  mystery  \
0     0.000000  0.482990       0.0  ...   0.0      0.0      0.0   
1     0.000000  0.593662       0.0  ...   0.0      0.0      0.0   
2     0.000000  0.000000       0.0  ...   0.0      0.0      0.0   
3     0.466405  0.000000       0.0  ...   0.0      0.0      0.0   
4     0.000000  0.000000       0.0  ...   0.0      0.0      0.0   
...        ...       ...       ...  ...   ...      ...      ...   
9737  0.000000  0.575034       0.0  ...   0.0      0.0      0.0   
9738  0.000000  0.638968       0.0  ...   0.0      0.0      0.0   
9739  1.000000  0.000000       0.0  ...   0.0      0.0      0.0   
9740  0.000000  0.000000       0.0  ...   0.0      0.0      0.0   
9741  0.000000  0.000000       0.0  ...   0.0      0.0      0.0   

      nogenreslisted   romance  scifi  thriller  war  western  movieId  
0                0.0  0.000000    0.0       0.0  0.0      0.0        1  
1                0.0  0.000000    0.0       0.0  0.0      0.0        2  
2                0.0  0.821009    0.0       0.0  0.0      0.0        3  
3                0.0  0.726241    0.0       0.0  0.0      0.0        4  
4                0.0  0.000000    0.0       0.0  0.0      0.0        5  
...              ...       ...    ...       ...  ...      ...      ...  
9737             0.0  0.000000    0.0       0.0  0.0      0.0   193581  
9738             0.0  0.000000    0.0       0.0  0.0      0.0   193583  
9739             0.0  0.000000    0.0       0.0  0.0      0.0   193585  
9740             0.0  0.000000    0.0       0.0  0.0      0.0   193587  
9741             0.0  0.000000    0.0       0.0  0.0      0.0   193609  

[9742 rows x 21 columns]

Делаем вектор из тегов:

In [83]:
tags

userId  movieId               tag   timestamp
0          2    60756             funny  1445714994
1          2    60756   Highly quotable  1445714996
2          2    60756      will ferrell  1445714992
3          2    89774      Boxing story  1445715207
4          2    89774               MMA  1445715200
...      ...      ...               ...         ...
3678     606     7382         for katie  1171234019
3679     606     7936           austere  1173392334
3680     610     3265            gun fu  1493843984
3681     610     3265  heroic bloodshed  1493843978
3682     610   168248  Heroic Bloodshed  1493844270

[3683 rows x 4 columns]

Объединяем таблицы:

In [95]:
movies_with_tags = movies.merge(tags, on='movieId')
movies_with_tags.head(10)

movieId                               title  \
0        1                    Toy Story (1995)   
1        1                    Toy Story (1995)   
2        1                    Toy Story (1995)   
3        2                      Jumanji (1995)   
4        2                      Jumanji (1995)   
5        2                      Jumanji (1995)   
6        2                      Jumanji (1995)   
7        3             Grumpier Old Men (1995)   
8        3             Grumpier Old Men (1995)   
9        5  Father of the Bride Part II (1995)   

                                        genres  userId               tag  \
0  Adventure|Animation|Children|Comedy|Fantasy     336             pixar   
1  Adventure|Animation|Children|Comedy|Fantasy     474             pixar   
2  Adventure|Animation|Children|Comedy|Fantasy     567               fun   
3                   Adventure|Children|Fantasy      62           fantasy   
4                   Adventure|Children|Fantasy      62  magic board game   
5                   Adventure|Children|Fantasy      62    Robin Williams   
6                   Adventure|Children|Fantasy     474              game   
7                               Comedy|Romance     289             moldy   
8                               Comedy|Romance     289               old   
9                                       Comedy     474         pregnancy   

    timestamp  
0  1139045764  
1  1137206825  
2  1525286013  
3  1528843929  
4  1528843932  
5  1528843907  
6  1137375552  
7  1143424860  
8  1143424860  
9  1137373903

In [89]:
def change_string(s):
    return str(s).replace(' ', '').replace('-', '').lower()

tag_strings = []
movieId = []


for movieid, group in tqdm(movies_with_tags.groupby('movieId')):
    tag_strings.append(' '.join([change_string(s) for s in group.tag.values]))
#    print(movieid)
    movieId.append(movieid)

  0%|          | 0/1572 [00:00<?, ?it/s]

In [90]:
tag_strings[:10]

['pixar pixar fun',
 'fantasy magicboardgame robinwilliams game',
 'moldy old',
 'pregnancy remake',
 'remake',
 'politics president',
 'politics president',
 'mafia',
 'janeausten',
 'hollywood']

In [92]:
movieId[:10]

[1, 2, 3, 5, 7, 11, 14, 16, 17, 21]

In [96]:
tfidf_tag = TfidfVectorizer()
X_train_tfidf_tag = tfidf_tag.fit_transform(tag_strings)
X_train_tfidf_tag

<1572x1472 sparse matrix of type '<class 'numpy.float64'>'
	with 3598 stored elements in Compressed Sparse Row format>

In [98]:
tags_idf = pd.DataFrame(X_train_tfidf_tag.toarray(), columns=tfidf_tag.get_feature_names_out())
tags_idf

06oscarnominatedbestmovieanimation  1900s  1920s  1950s  1960s  1970s  \
0                                    0.0    0.0    0.0    0.0    0.0    0.0   
1                                    0.0    0.0    0.0    0.0    0.0    0.0   
2                                    0.0    0.0    0.0    0.0    0.0    0.0   
3                                    0.0    0.0    0.0    0.0    0.0    0.0   
4                                    0.0    0.0    0.0    0.0    0.0    0.0   
...                                  ...    ...    ...    ...    ...    ...   
1567                                 0.0    0.0    0.0    0.0    0.0    0.0   
1568                                 0.0    0.0    0.0    0.0    0.0    0.0   
1569                                 0.0    0.0    0.0    0.0    0.0    0.0   
1570                                 0.0    0.0    0.0    0.0    0.0    0.0   
1571                                 0.0    0.0    0.0    0.0    0.0    0.0   

      1980s  1990s  2001like  2danimation  ...  worldwari  worldwarii  \
0       0.0    0.0       0.0          0.0  ...        0.0         0.0   
1       0.0    0.0       0.0          0.0  ...        0.0         0.0   
2       0.0    0.0       0.0          0.0  ...        0.0         0.0   
3       0.0    0.0       0.0          0.0  ...        0.0         0.0   
4       0.0    0.0       0.0          0.0  ...        0.0         0.0   
...     ...    ...       ...          ...  ...        ...         ...   
1567    0.0    0.0       0.0          0.0  ...        0.0         0.0   
1568    0.0    0.0       0.0          0.0  ...        0.0         0.0   
1569    0.0    0.0       0.0          0.0  ...        0.0         0.0   
1570    0.0    0.0       0.0          0.0  ...        0.0         0.0   
1571    0.0    0.0       0.0          0.0  ...        0.0         0.0   

      writing  wrongfulimprisonment  wry  youngermen  zither  zoekazan  \
0         0.0                   0.0  0.0         0.0     0.0       0.0   
1         0.0                   0.0  0.0         0.0     0.0       0.0   
2         0.0                   0.0  0.0         0.0     0.0       0.0   
3         0.0                   0.0  0.0         0.0     0.0       0.0   
4         0.0                   0.0  0.0         0.0     0.0       0.0   
...       ...                   ...  ...         ...     ...       ...   
1567      0.0                   0.0  0.0         0.0     0.0       0.0   
1568      0.0                   0.0  0.0         0.0     0.0       0.0   
1569      0.0                   0.0  0.0         0.0     0.0       0.0   
1570      0.0                   0.0  0.0         0.0     0.0       0.0   
1571      0.0                   0.0  0.0         0.0     0.0       0.0   

      zombies  zooeydeschanel  
0         0.0             0.0  
1         0.0             0.0  
2         0.0             0.0  
3         0.0             0.0  
4         0.0             0.0  
...       ...             ...  
1567      0.0             0.0  
1568      0.0             0.0  
1569      0.0             0.0  
1570      0.0             0.0  
1571      0.0             0.0  

[1572 rows x 1472 columns]

Теперь добавляем столбец с id фильмов:

In [99]:
tags_idf['movieId'] = movieId
tags_idf

06oscarnominatedbestmovieanimation  1900s  1920s  1950s  1960s  1970s  \
0                                    0.0    0.0    0.0    0.0    0.0    0.0   
1                                    0.0    0.0    0.0    0.0    0.0    0.0   
2                                    0.0    0.0    0.0    0.0    0.0    0.0   
3                                    0.0    0.0    0.0    0.0    0.0    0.0   
4                                    0.0    0.0    0.0    0.0    0.0    0.0   
...                                  ...    ...    ...    ...    ...    ...   
1567                                 0.0    0.0    0.0    0.0    0.0    0.0   
1568                                 0.0    0.0    0.0    0.0    0.0    0.0   
1569                                 0.0    0.0    0.0    0.0    0.0    0.0   
1570                                 0.0    0.0    0.0    0.0    0.0    0.0   
1571                                 0.0    0.0    0.0    0.0    0.0    0.0   

      1980s  1990s  2001like  2danimation  ...  worldwarii  writing  \
0       0.0    0.0       0.0          0.0  ...         0.0      0.0   
1       0.0    0.0       0.0          0.0  ...         0.0      0.0   
2       0.0    0.0       0.0          0.0  ...         0.0      0.0   
3       0.0    0.0       0.0          0.0  ...         0.0      0.0   
4       0.0    0.0       0.0          0.0  ...         0.0      0.0   
...     ...    ...       ...          ...  ...         ...      ...   
1567    0.0    0.0       0.0          0.0  ...         0.0      0.0   
1568    0.0    0.0       0.0          0.0  ...         0.0      0.0   
1569    0.0    0.0       0.0          0.0  ...         0.0      0.0   
1570    0.0    0.0       0.0          0.0  ...         0.0      0.0   
1571    0.0    0.0       0.0          0.0  ...         0.0      0.0   

      wrongfulimprisonment  wry  youngermen  zither  zoekazan  zombies  \
0                      0.0  0.0         0.0     0.0       0.0      0.0   
1                      0.0  0.0         0.0     0.0       0.0      0.0   
2                      0.0  0.0         0.0     0.0       0.0      0.0   
3                      0.0  0.0         0.0     0.0       0.0      0.0   
4                      0.0  0.0         0.0     0.0       0.0      0.0   
...                    ...  ...         ...     ...       ...      ...   
1567                   0.0  0.0         0.0     0.0       0.0      0.0   
1568                   0.0  0.0         0.0     0.0       0.0      0.0   
1569                   0.0  0.0         0.0     0.0       0.0      0.0   
1570                   0.0  0.0         0.0     0.0       0.0      0.0   
1571                   0.0  0.0         0.0     0.0       0.0      0.0   

      zooeydeschanel  movieId  
0                0.0        1  
1                0.0        2  
2                0.0        3  
3                0.0        5  
4                0.0        7  
...              ...      ...  
1567             0.0   183611  
1568             0.0   184471  
1569             0.0   187593  
1570             0.0   187595  
1571             0.0   193565  

[1572 rows x 1473 columns]

Объединяем таблицы с tf-idf:

In [101]:
super_table = movie_genres_table_id.merge(tags_idf, on='movieId', how='outer')
super_table

action_x  adventure_x  animation_x  children_x  comedy_x  crime_x  \
0     0.000000     0.416846     0.516225    0.504845  0.267586      0.0   
1     0.000000     0.512361     0.000000    0.620525  0.000000      0.0   
2     0.000000     0.000000     0.000000    0.000000  0.570915      0.0   
3     0.000000     0.000000     0.000000    0.000000  0.505015      0.0   
4     0.000000     0.000000     0.000000    0.000000  1.000000      0.0   
...        ...          ...          ...         ...       ...      ...   
9737  0.436010     0.000000     0.614603    0.000000  0.318581      0.0   
9738  0.000000     0.000000     0.682937    0.000000  0.354002      0.0   
9739  0.000000     0.000000     0.000000    0.000000  0.000000      0.0   
9740  0.578606     0.000000     0.815607    0.000000  0.000000      0.0   
9741  0.000000     0.000000     0.000000    0.000000  1.000000      0.0   

      documentary_x   drama_x  fantasy_x  filmnoir_x  ...  worldwari  \
0               0.0  0.000000   0.482990         0.0  ...        0.0   
1               0.0  0.000000   0.593662         0.0  ...        0.0   
2               0.0  0.000000   0.000000         0.0  ...        0.0   
3               0.0  0.466405   0.000000         0.0  ...        NaN   
4               0.0  0.000000   0.000000         0.0  ...        0.0   
...             ...       ...        ...         ...  ...        ...   
9737            0.0  0.000000   0.575034         0.0  ...        NaN   
9738            0.0  0.000000   0.638968         0.0  ...        NaN   
9739            0.0  1.000000   0.000000         0.0  ...        NaN   
9740            0.0  0.000000   0.000000         0.0  ...        NaN   
9741            0.0  0.000000   0.000000         0.0  ...        NaN   

      worldwarii  writing  wrongfulimprisonment  wry  youngermen  zither  \
0            0.0      0.0                   0.0  0.0         0.0     0.0   
1            0.0      0.0                   0.0  0.0         0.0     0.0   
2            0.0      0.0                   0.0  0.0         0.0     0.0   
3            NaN      NaN                   NaN  NaN         NaN     NaN   
4            0.0      0.0                   0.0  0.0         0.0     0.0   
...          ...      ...                   ...  ...         ...     ...   
9737         NaN      NaN                   NaN  NaN         NaN     NaN   
9738         NaN      NaN                   NaN  NaN         NaN     NaN   
9739         NaN      NaN                   NaN  NaN         NaN     NaN   
9740         NaN      NaN                   NaN  NaN         NaN     NaN   
9741         NaN      NaN                   NaN  NaN         NaN     NaN   

      zoekazan  zombies  zooeydeschanel  
0          0.0      0.0             0.0  
1          0.0      0.0             0.0  
2          0.0      0.0             0.0  
3          NaN      NaN             NaN  
4          0.0      0.0             0.0  
...        ...      ...             ...  
9737       NaN      NaN             NaN  
9738       NaN      NaN             NaN  
9739       NaN      NaN             NaN  
9740       NaN      NaN             NaN  
9741       NaN      NaN             NaN  

[9742 rows x 1493 columns]

Заменяем отсутствие значений на нули:

In [103]:
super_table = super_table.fillna(0)
super_table

action_x  adventure_x  animation_x  children_x  comedy_x  crime_x  \
0     0.000000     0.416846     0.516225    0.504845  0.267586      0.0   
1     0.000000     0.512361     0.000000    0.620525  0.000000      0.0   
2     0.000000     0.000000     0.000000    0.000000  0.570915      0.0   
3     0.000000     0.000000     0.000000    0.000000  0.505015      0.0   
4     0.000000     0.000000     0.000000    0.000000  1.000000      0.0   
...        ...          ...          ...         ...       ...      ...   
9737  0.436010     0.000000     0.614603    0.000000  0.318581      0.0   
9738  0.000000     0.000000     0.682937    0.000000  0.354002      0.0   
9739  0.000000     0.000000     0.000000    0.000000  0.000000      0.0   
9740  0.578606     0.000000     0.815607    0.000000  0.000000      0.0   
9741  0.000000     0.000000     0.000000    0.000000  1.000000      0.0   

      documentary_x   drama_x  fantasy_x  filmnoir_x  ...  worldwari  \
0               0.0  0.000000   0.482990         0.0  ...        0.0   
1               0.0  0.000000   0.593662         0.0  ...        0.0   
2               0.0  0.000000   0.000000         0.0  ...        0.0   
3               0.0  0.466405   0.000000         0.0  ...        0.0   
4               0.0  0.000000   0.000000         0.0  ...        0.0   
...             ...       ...        ...         ...  ...        ...   
9737            0.0  0.000000   0.575034         0.0  ...        0.0   
9738            0.0  0.000000   0.638968         0.0  ...        0.0   
9739            0.0  1.000000   0.000000         0.0  ...        0.0   
9740            0.0  0.000000   0.000000         0.0  ...        0.0   
9741            0.0  0.000000   0.000000         0.0  ...        0.0   

      worldwarii  writing  wrongfulimprisonment  wry  youngermen  zither  \
0            0.0      0.0                   0.0  0.0         0.0     0.0   
1            0.0      0.0                   0.0  0.0         0.0     0.0   
2            0.0      0.0                   0.0  0.0         0.0     0.0   
3            0.0      0.0                   0.0  0.0         0.0     0.0   
4            0.0      0.0                   0.0  0.0         0.0     0.0   
...          ...      ...                   ...  ...         ...     ...   
9737         0.0      0.0                   0.0  0.0         0.0     0.0   
9738         0.0      0.0                   0.0  0.0         0.0     0.0   
9739         0.0      0.0                   0.0  0.0         0.0     0.0   
9740         0.0      0.0                   0.0  0.0         0.0     0.0   
9741         0.0      0.0                   0.0  0.0         0.0     0.0   

      zoekazan  zombies  zooeydeschanel  
0          0.0      0.0             0.0  
1          0.0      0.0             0.0  
2          0.0      0.0             0.0  
3          0.0      0.0             0.0  
4          0.0      0.0             0.0  
...        ...      ...             ...  
9737       0.0      0.0             0.0  
9738       0.0      0.0             0.0  
9739       0.0      0.0             0.0  
9740       0.0      0.0             0.0  
9741       0.0      0.0             0.0  

[9742 rows x 1493 columns]

Теперь вытащим ещё оценки пользователей, то, что нам надо предсказать:

In [105]:
ratings

userId  movieId  rating   timestamp
0            1        1     4.0   964982703
1            1        3     4.0   964981247
2            1        6     4.0   964982224
3            1       47     5.0   964983815
4            1       50     5.0   964982931
...        ...      ...     ...         ...
100831     610   166534     4.0  1493848402
100832     610   168248     5.0  1493850091
100833     610   168250     5.0  1494273047
100834     610   168252     5.0  1493846352
100835     610   170875     3.0  1493846415

[100836 rows x 4 columns]

Сразу посчитаем медиану для одного фильма:

In [107]:
rating_median = ratings.groupby(['movieId'])['rating']. median().reset_index()
rating_median

movieId  rating
0           1     4.0
1           2     3.5
2           3     3.0
3           4     3.0
4           5     3.0
...       ...     ...
9719   193581     4.0
9720   193583     3.5
9721   193585     3.5
9722   193587     3.5
9723   193609     4.0

[9724 rows x 2 columns]

Слепим с получившейся таблицей idf:

In [127]:
super_table_2 = rating_median.merge(super_table, on='movieId')
super_table_2

movieId  rating  action_x  adventure_x  animation_x  children_x  \
0           1     4.0  0.000000     0.416846     0.516225    0.504845   
1           2     3.5  0.000000     0.512361     0.000000    0.620525   
2           3     3.0  0.000000     0.000000     0.000000    0.000000   
3           4     3.0  0.000000     0.000000     0.000000    0.000000   
4           5     3.0  0.000000     0.000000     0.000000    0.000000   
...       ...     ...       ...          ...          ...         ...   
9719   193581     4.0  0.436010     0.000000     0.614603    0.000000   
9720   193583     3.5  0.000000     0.000000     0.682937    0.000000   
9721   193585     3.5  0.000000     0.000000     0.000000    0.000000   
9722   193587     3.5  0.578606     0.000000     0.815607    0.000000   
9723   193609     4.0  0.000000     0.000000     0.000000    0.000000   

      comedy_x  crime_x  documentary_x   drama_x  ...  worldwari  worldwarii  \
0     0.267586      0.0            0.0  0.000000  ...        0.0         0.0   
1     0.000000      0.0            0.0  0.000000  ...        0.0         0.0   
2     0.570915      0.0            0.0  0.000000  ...        0.0         0.0   
3     0.505015      0.0            0.0  0.466405  ...        0.0         0.0   
4     1.000000      0.0            0.0  0.000000  ...        0.0         0.0   
...        ...      ...            ...       ...  ...        ...         ...   
9719  0.318581      0.0            0.0  0.000000  ...        0.0         0.0   
9720  0.354002      0.0            0.0  0.000000  ...        0.0         0.0   
9721  0.000000      0.0            0.0  1.000000  ...        0.0         0.0   
9722  0.000000      0.0            0.0  0.000000  ...        0.0         0.0   
9723  1.000000      0.0            0.0  0.000000  ...        0.0         0.0   

      writing  wrongfulimprisonment  wry  youngermen  zither  zoekazan  \
0         0.0                   0.0  0.0         0.0     0.0       0.0   
1         0.0                   0.0  0.0         0.0     0.0       0.0   
2         0.0                   0.0  0.0         0.0     0.0       0.0   
3         0.0                   0.0  0.0         0.0     0.0       0.0   
4         0.0                   0.0  0.0         0.0     0.0       0.0   
...       ...                   ...  ...         ...     ...       ...   
9719      0.0                   0.0  0.0         0.0     0.0       0.0   
9720      0.0                   0.0  0.0         0.0     0.0       0.0   
9721      0.0                   0.0  0.0         0.0     0.0       0.0   
9722      0.0                   0.0  0.0         0.0     0.0       0.0   
9723      0.0                   0.0  0.0         0.0     0.0       0.0   

      zombies  zooeydeschanel  
0         0.0             0.0  
1         0.0             0.0  
2         0.0             0.0  
3         0.0             0.0  
4         0.0             0.0  
...       ...             ...  
9719      0.0             0.0  
9720      0.0             0.0  
9721      0.0             0.0  
9722      0.0             0.0  
9723      0.0             0.0  

[9724 rows x 1494 columns]

Не очень понимаю, зачем использовать данные в целом об оценках именно пользователя, если у нас оценка рейтинга фильмов. Ну и как их прилепить к этой таблице, даже если использовать. 

Попробуем предсказать без данных о пользователях:

In [128]:
X = super_table_2

In [130]:
y = X[['rating']]

In [131]:
del X['rating']

In [132]:
del X['movieId']

In [133]:
#Разделите выборку на обучающее и тестовое подмножество. 80% данных оставить на обучающее множество, 20% на тестовое.
X_new = X

In [134]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Использую вариацию линейной регрессии для задач регрессии. 

Метод регрессии лассо (LASSO, Least Absolute Shrinkage and Selection Operator) — это вариация линейной регрессии, специально адаптированная для данных, которые демонстрируют сильную мультиколлинеарность (то есть сильную корреляцию признаков друг с другом)

In [138]:
from sklearn.linear_model import LassoCV
model = LassoCV()
model.fit(X_train, y_train)

C:\Users\Алехандро\AppData\Roaming\Python\Python311\site-packages\sklearn\linear_model\_coordinate_descent.py:1563: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LassoCV()

In [139]:
model.fit(X_train, y_train)
predictions = model.predict(X_test)

C:\Users\Алехандро\AppData\Roaming\Python\Python311\site-packages\sklearn\linear_model\_coordinate_descent.py:1563: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [148]:
predictions

array([3.56347804, 3.31884977, 2.85446617, ..., 3.0062264 , 3.92129414,
       3.37730373])

In [150]:
y_test

rating
1145    4.00
8505    4.00
3903    5.00
1608    3.25
9141    3.75
...      ...
7676    4.00
1918    3.25
8728    3.75
4184    4.00
5955    3.50

[1945 rows x 1 columns]

In [142]:
from sklearn.metrics import mean_squared_error
from math import sqrt

In [152]:
sqrt(mean_squared_error(y_test, predictions)) 

0.8212149622668329

### Вывод:
Средняя квадратичная ошибка 0.82. То есть предсказанные значения в среднем отличаются от реальных меньше чем на 1. Считаю, что это достаточно неплохой результат, если учесть, что в рейтинге пятибальная шкала оценивания фильмов.